specs/latest/1.0/index.html

```diff
@@ -1979,6 +1980,20 @@ <h4>Attributes</h4>
             The actual height of the drawing buffer. May be different from the
             <code>height</code> attribute of the <code>HTMLCanvasElement</code> if
             the implementation is unable to satisfy the requested width or height.
+        <dt>
+            <code class=attribute-name>
+                <a id="DOM-WebGLRenderingContext-colorSpace">
+                    colorSpace
+                </a>
+            </code>
+            of type <code><a href="https://html.spec.whatwg.org/multipage/canvas.html#predefinedcolorspace">PredefinedColorSpace</a></code>
+            <a href="#refsPREDEFINEDCOLORSPACE">(specification)</a>
+        <dd>
+            The color space in which the drawing buffer's contents are assumed to
+            have been produced. Changing this attribute affects the on-screen display
+            of the rendering context, but does not change the drawing buffer's
+            contents. This attribute also affects how <code>TexImageSource</code>
+            sources are uploaded to textures.
```

Comment:
It's kind of surprising to me that this attribute changes upload behavior as well. In that way, it's one knob that's changing both the canvas behavior (for display) and the GL context behavior (for upload). In the latter case it's GL-style global state, where that state affects the behavior of particular GL calls. A usage pattern I would expect to emerge is that applications do things like:

```js
const oldColorSpace = gl.colorSpace;
gl.colorSpace = 'srgb';
// upload an sRGB JPG; keep it in sRGB (rather than converting it now), for internal processing by the app
gl.texImage2D(/* ... */);
gl.colorSpace = oldColorSpace; // set the color space back to the previous value for canvas output
```

In this particular example the app would probably be more likely to set `UNPACK_COLORSPACE_CONVERSION_WEBGL` to `NONE` instead, but I think it illustrates the point: can we have `gl.colorSpace` only affect the canvas output, and have a separate GL state for input? (`UNPACK_COLORSPACE_WEBGL`)

The shape of this API seems to imply that `colorSpace` sets the "working color space" for the entire WebGL context. But since it's mutable, that interpretation probably doesn't make sense. I think that interpretation makes sense in 2d canvas - if I'm reading the HTML spec correctly (please verify), the 2d canvas color space cannot change. But in WebGL, color spaces only matter at input and output points, so there should be no "working color space".

Reply:
These are fair points. @jdashg, @ccameron-chromium and I discussed this, and our opinion was that it would be most useful to WebGL applications to be able to set the `colorSpace` on the fly - especially given #3222, which will allow reallocation of the drawing buffer with a specific format. The use case differs from the 2D rendering context, where the canvas is fully color managed - that's why the color space is fixed at context creation time in that API. In WebGL, color management is basically up to the application, and feature detection is needed to understand the supported drawing buffer formats, so `colorSpace` needs to be mutable.

Since the `colorSpace` is mutable in WebGL, and it will be inexpensive to change it, it seemed unnecessary to specify new pixel unpack state for the color space of incoming texture data. It would also be a bit confusing, since it's expected that nearly all applications that would change this setting would want their texture data converted to the canvas's working color space at the time of upload.

Given this background, what do you think? Note, I'm pushing a revision to this spec based on your feedback above and hope this conversation doesn't get lost.

Reply:
I do agree the colorSpace should be mutable. But to me, as written, it _looks_ like a "working" color space, when it is really just a knob that controls both input and output from WebGL. 

> It would also be a bit confusing, since it's expected that nearly all applications that would change this setting would want their texture data converted to the canvas's working color space at the time of upload.

I'm still not sure about this, although I don't have a specific, sufficiently-common counterexample in mind. To me it seems less confusing if the two things are controlled by two knobs instead of by one that you have to set and reset.

Most examples I can think of with more advanced usage would just use `UNPACK_COLORSPACE_CONVERSION_WEBGL` `NONE`. Here's the best one I have: a 3D application with a linear physical rendering pipeline wants to output with `colorSpace = 'linear-srgb'`, but upload assets, with conversion to `'srgb'`, into an SRGB-encoded texture format.

Reply:
The new text's been written to try to make it clear that changing `colorSpace` has no effect on the contents of the drawing buffer. Would a rewording make it clearer?

Regarding the texture upload behavior: consider the case of a WebGL image viewer application using the `display-p3` color space. The application author will just want all incoming assets converted by the browser into that color space. That will be the behavior with this PR. @jdashg , @ccameron-chromium and I couldn't think of many (or any) situations where you'd want the browser to convert incoming assets to a different color space than the one being used for the canvas's display. For this reason I think saving and restoring the `colorSpace` property will be rare.

`UNPACK_COLORSPACE_CONVERSION_WEBGL` has limited applicability; it only works for HTMLImageElements, and ImageBitmap created from a Blob is a better alternative for that path nowadays (in Chrome, at least - Firefox engineers will disagree). It has no effect for video elements, for example.  This is another reason we didn't want to add a new pixel unpack state to specify a destination color space for texture uploads.

Reply:
I think the wording is clear, it's more about the clarity of the API surface. People are going to need documentation and experimentation anyway though, so as long as it's consistent and well tested, it's probably not that important. I'm not that opposed to the API as written, so if others prefer it this way, it seems OK.

I forgot `UNPACK_COLORSPACE_CONVERSION_WEBGL` wouldn't consistently apply in the same paths that a `UNPACK_COLORSPACE_WEBGL` would. That indeed would be confusing.

Reply:
Having had a bit more discussion about this, I think we don't want to expose a color space attribute in ImageBitmapOptions. A few reasons

1. Just adding a color space would still be under-specified, because we still want to specify a pixel format. E.g, an application can want to import an P3 image as srgb-linear float16. If we default to converting to 8-bit, this would clamp and quantize results inappropriately. If we were to add a pixel format, this would have to be an ImageBitmap-specific pixel format, with a compatibility matrix for WebGL and WebGPU.

2. Converting color space in ImageBitmap is probably not the most efficient path for importing to WebGL. The most efficient format, for JPEG among other formats, is to 
* decode to YUV (most CPU-efficient, b/c no need for YUV->RGB conversion)
* upload YUV (most efficient, because is smaller upload size than RGB)
* convert YUV->RGB and do color space conversion while importing to WebGL
The color conversion is almost-for-free compared to just doing the YUV->RGB conversion, and is done on the GPU where it's most efficient.

3. If we want move various steps (e.g, the CPU->GPU upload) to be done asynchronously, we would need to add to ImageBitmapOptions a parameter specifying the target device and the target API (e.g, send to WebGL on GPU 0).

Given all of that, I think it's best to keep the image in as close to its "native" format for as long as possible. It may be possible to do an eager upload to the GPU (when we know which GPU that will be).

Reply:
All good points. Have changed this PR to have unpackColorSpace apply to ImageBitmaps.